File: oss/clawpipe/dotnet-sdk/ClawPipe/Cache.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;
using System.Text;

namespace ClawPipe;

/// <summary>
/// In-memory prompt cache with SHA-256 keys, TTL expiry, and LRU-style eviction.
/// Thread-safe via ConcurrentDictionary.
/// </summary>
public sealed class Cache
{
    private readonly ConcurrentDictionary<string, CacheEntry> _store = new();
    private readonly TimeSpan _defaultTtl;
    private readonly int _maxEntries;
    private long _totalHits;
    private long _totalMisses;

    public Cache(TimeSpan? defaultTtl = null, int maxEntries = 10_000)
    {
        _defaultTtl = defaultTtl ?? TimeSpan.FromMinutes(5);
        _maxEntries = maxEntries;
    }

    /// <summary>Compute a SHA-256 cache key from a prompt string.</summary>
    public static string ComputeKey(string prompt)
    {
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(prompt));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    /// <summary>Retrieve a cached value. Returns null if missing or expired.</summary>
    public string? Get(string prompt)
    {
        var key = ComputeKey(prompt);
        if (!_store.TryGetValue(key, out var entry))
        {
            Interlocked.Increment(ref _totalMisses);
            return null;
        }

        if (DateTime.UtcNow > entry.Expiry)
        {
            _store.TryRemove(key, out _);
            Interlocked.Increment(ref _totalMisses);
            return null;
        }

        entry.IncrementHits();
        Interlocked.Increment(ref _totalHits);
        return entry.Value;
    }

    /// <summary>Store a value in cache with optional per-entry TTL override.</summary>
    public void Set(string prompt, string response, TimeSpan? ttl = null)
    {
        EvictIfFull();
        var key = ComputeKey(prompt);
        var expiry = DateTime.UtcNow + (ttl ?? _defaultTtl);
        _store[key] = new CacheEntry(response, expiry);
    }

    /// <summary>Remove a specific prompt from cache.</summary>
    public bool Delete(string prompt) => _store.TryRemove(ComputeKey(prompt), out _);

    /// <summary>Clear all cached entries and reset statistics.</summary>
    public void Clear()
    {
        _store.Clear();
        Interlocked.Exchange(ref _totalHits, 0);
        Interlocked.Exchange(ref _totalMisses, 0);
    }

    /// <summary>Remove all expired entries. Returns number removed.</summary>
    public int Prune()
    {
        var now = DateTime.UtcNow;
        var expired = _store.Where(kv => now > kv.Value.Expiry).Select(kv => kv.Key).ToList();
        foreach (var k in expired) _store.TryRemove(k, out _);
        return expired.Count;
    }

    /// <summary>Cache performance statistics.</summary>
    public CacheStats Stats()
    {
        var hits = Interlocked.Read(ref _totalHits);
        var misses = Interlocked.Read(ref _totalMisses);
        var total = hits + misses;
        var rate = total > 0 ? (double)hits / total * 100 : 0;
        return new CacheStats(_store.Count, hits, misses, $"{rate:F1}%");
    }

    private void EvictIfFull()
    {
        if (_store.Count < _maxEntries) return;
        var toRemove = (int)Math.Ceiling(_maxEntries * 0.1);
        var leastHit = _store.OrderBy(kv => kv.Value.Hits).Take(toRemove).Select(kv => kv.Key);
        foreach (var k in leastHit) _store.TryRemove(k, out _);
    }
}

/// <summary>A single entry in the cache.</summary>
internal sealed class CacheEntry(string value, DateTime expiry)
{
    public string Value { get; } = value;
    public DateTime Expiry { get; } = expiry;
    private long _hits;
    public long Hits => Interlocked.Read(ref _hits);
    public void IncrementHits() => Interlocked.Increment(ref _hits);
}

/// <summary>Cache performance statistics snapshot.</summary>
public sealed record CacheStats(long Size, long Hits, long Misses, string HitRate);

File: oss/clawpipe/php-sdk/src/ClawPipe/Cache.php
<?php

declare(strict_types=1);

namespace ClawPipe;

/**
 * In-memory prompt cache with SHA-256 keying and TTL expiry.
 */
class Cache
{
    /** @var array<string, array{value: string, expires_at: int}> */
    private array $store = [];

    private int $hits   = 0;
    private int $misses = 0;

    /**
     * @param int $defaultTtl Default TTL in seconds (default 300 = 5 min).
     * @param int $maxEntries Maximum entries before LRU eviction.
     */
    public function __construct(
        private int $defaultTtl  = 300,
        private int $maxEntries  = 10_000,
    ) {}

    /**
     * Retrieve a cached response for a prompt.
     *
     * @param string $prompt The original prompt text.
     * @return string|null Cached response, or null on miss / expiry.
     */
    public function get(string $prompt): ?string
    {
        $key   = $this->hashKey($prompt);
        $entry = $this->store[$key] ?? null;

        if ($entry === null) {
            $this->misses++;
            return null;
        }

        if (time() >= $entry['expires_at']) {
            unset($this->store[$key]);
            $this->misses++;
            return null;
        }

        $this->hits++;
        return $entry['value'];
    }

    /**
     * Store a response in the cache.
     *
     * @param string $prompt   The original prompt text used as cache key.
     * @param string $response The LLM response to cache.
     * @param int    $ttl      TTL in seconds (0 = use default).
     */
    public function set(string $prompt, string $response, int $ttl = 0): void
    {
        $this->evictIfFull();

        $key = $this->hashKey($prompt);
        $this->store[$key] = [
            'value'      => $response,
            'expires_at' => time() + ($ttl > 0 ? $ttl : $this->defaultTtl),
        ];
    }

    /**
     * Check whether a valid (non-expired) entry exists.
     */
    public function has(string $prompt): bool
    {
        return $this->get($prompt) !== null;
    }

    /**
     * Remove a specific cached prompt.
     */
    public function delete(string $prompt): bool
    {
        $key = $this->hashKey($prompt);
        if (isset($this->store[$key])) {
            unset($this->store[$key]);
            return true;
        }
        return false;
    }

    /**
     * Flush the entire cache and reset statistics.
     */
    public function clear(): void
    {
        $this->store  = [];
        $this->hits   = 0;
        $this->misses = 0;
    }

    /**
     * Remove all expired entries and return count removed.
     */
    public function prune(): int
    {
        $now     = time();
        $removed = 0;
        foreach (array_keys($this->store) as $key) {
            if ($now >= $this->store[$key]['expires_at']) {
                unset($this->store[$key]);
                $removed++;
            }
        }
        return $removed;
    }

    /**
     * Return cache performance statistics.
     *
     * @return array{size: int, hits: int, misses: int, hit_rate: string}
     */
    public function stats(): array
    {
        $total   = $this->hits + $this->misses;
        $hitRate = $total > 0 ? round($this->hits / $total * 100, 1) . '%' : '0.0%';

        return [
            'size'     => count($this->store),
            'hits'     => $this->hits,
            'misses'   => $this->misses,
            'hit_rate' => $hitRate,
        ];
    }

    /**
     * SHA-256 hash of the prompt, prefixed with "cp_".
     */
    private function hashKey(string $prompt): string
    {
        return 'cp_' . hash('sha256', $prompt);
    }

    /**
     * Evict expired entries first; if still full, drop oldest by expiry.
     */
    private function evictIfFull(): void
    {
        if (count($this->store) < $this->maxEntries) {
            return;
        }

        // First pass: remove expired
        $this->prune();

        if (count($this->store) < $this->maxEntries) {
            return;
        }

        // Second pass: evict 10% oldest entries
        uasort($this->store, fn($a, $b) => $a['expires_at'] <=> $b['expires_at']);
        $toRemove = (int) ceil($this->maxEntries * 0.1);
        $keys     = array_keys($this->store);
        for ($i = 0; $i < min($toRemove, count($keys)); $i++) {
            unset($this->store[$keys[$i]]);
        }
    }
}
